File: func-pipes/static/js/power2/nodes/decision.js
/*
  nodes/decision.js — Decision / Router Node
  ─────────────────────────────────────────────────────────────────────────────
  Routes an inbound signal to one (or many) of its N outbound outputs based on
  a user-defined `decide()` function evaluated every tick and on every signal
  arrival.

  Designed for clean subclassing:

      class VoltageRouter extends DecisionNode {
          static type        = 'volt-router'
          static label       = 'Voltage Router'
          static outputCount = 3

          static defaults(id, preset = {}) {
              return { ...super.defaults(id, preset), minVolt: preset.minVolt ?? 180 }
          }

          // Return output index, array of indices (multicast), or null (block all).
          static decide(panel, signal) {
              if (!signal) return null
              if (signal.v >= 200) return 0
              if (signal.v >= panel.minVolt) return 1
              return 2
          }
      }

      NodeRegistry.register(VoltageRouter)

  Runtime callback (no subclass needed):

      const panel = graph.addType('decision')
      panel.decideCallback = (signal, panel) => signal?.v > 150 ? 0 : 1

  Extended state
  ──────────────
  outputCount     number    — how many outbound pips (set at creation via preset)
  tickInterval    number    — seconds between tick-driven re-evaluations (default 1.0)
  lastDecision    number|null — output index(es) chosen on last decide() call
  decideCallback  function|null — optional runtime override for decide()

  States: 'off' | 'routing' | 'blocked' | 'error'
*/

class DecisionNode extends NodeBase {

    static type        = 'decision'
    static label       = 'Decision'
    static group       = 'Control'
    static outputCount = 2              // default: 2 outputs; override in subclass

    static catalog = [
        { key: 'decision-2',   label: 'Decision (2 in → 2 out)', inputCount: 2, outputCount: 2 },
        { key: 'decision-2-3', label: 'Decision (2 in → 3 out)', inputCount: 2, outputCount: 3 },
        { key: 'decision-4',   label: 'Decision (4 in → 4 out)', inputCount: 4, outputCount: 4 },
    ]

    // ── State factory ─────────────────────────────────────────────────────────

    static defaults(id, preset = {}) {
        const inCount  = preset.inputCount  ?? this.outputCount   // default same as outputs
        const outCount = preset.outputCount ?? this.outputCount
        const base     = super.defaults(id, preset)
        return {
            ...base,
            // Configuration
            inputCount:     inCount,
            outputCount:    outCount,
            tickInterval:   preset.tickInterval ?? 1.0,
            // Runtime state
            lastDecision:   null,
            defaultOutput:  0,
            decideCallback: null,
            _tickAccum:     0,
            _lastInPip:     0,
            _pipSignals:    {},           // pipIndex → last signal received on that pip
            _isDecision:    true,
            // Override pips with correct counts
            pipsInbound:  Array.from({ length: inCount  }, (_, i) => ({ label: id, index: i })),
            pipsOutbound: Array.from({ length: outCount }, (_, i) => ({ label: id, index: i })),
        }
    }

    static configFields() {
        return [...super.configFields(), 'inputCount', 'outputCount', 'tickInterval', 'defaultOutput']
    }

    // ── Core hooks ────────────────────────────────────────────────────────────

    /**
     * Called whenever an upstream signal changes on any inbound pip.
     * `panel._lastInPip` is set by graph.receive() before this is called.
     */
    static apply(panel, signal, graph) {
        const inPip = panel._lastInPip ?? 0
        // Track per-pip signal so decide() can inspect all inputs
        panel._pipSignals[inPip] = signal
        this._route(panel, signal, graph, inPip)
    }

    /**
     * Periodic re-evaluation at `tickInterval` seconds.
     * Useful for time-based routing decisions that change independently of
     * signal value — e.g. "switch to output 1 after 10 seconds of inactivity".
     */
    static tick(panel, dt, graph) {
        if (panel.enabled === false) return
        panel._tickAccum += dt
        if (panel._tickAccum >= panel.tickInterval) {
            panel._tickAccum = 0
            // Re-route using the most recently active inbound pip
            this._route(panel, panel.signal, graph, panel._lastInPip ?? 0)
        }
    }

    static reset(panel, graph) {
        panel.lastDecision = null
        panel._tickAccum   = 0
        panel._pipSignals  = {}
        panel._lastInPip   = 0
        DecisionNode.dispatch(panel, 'decision:reset', {})
        super.reset(panel, graph)
        panel.pipsOutbound.forEach((_, i) => graph.emitTo(panel, i, null))
    }

    // ── Routing ───────────────────────────────────────────────────────────────

    /**
     * Evaluate decide(), then send the signal to the chosen output(s) and null
     * to all others.
     * @private
     */
    static _route(panel, signal, graph, inputIndex = 0) {
        let chosen
        try {
            chosen = this.decide(panel, signal, inputIndex, graph)
        } catch (err) {
            console.error(`[DecisionNode:${panel.id}] decide() threw:`, err)
            panel.state = 'error'
            DecisionNode.dispatch(panel, 'decision:error', { error: err.message })
            panel.pipsOutbound.forEach((_, i) => graph.emitTo(panel, i, null))
            return
        }

        const prev = panel.lastDecision
        panel.lastDecision = chosen

        if (chosen === null) {
            // Block all outputs
            panel.state = signal ? 'blocked' : 'off'
            panel.pipsOutbound.forEach((_, i) => graph.emitTo(panel, i, null))
            if (prev !== chosen)
                DecisionNode.dispatch(panel, 'decision:blocked', { prev, signal: !!signal })
            return
        }

        // Normalise to array to support multicast (return [0, 2] to emit to both)
        const targets = Array.isArray(chosen) ? chosen : [chosen]

        panel.pipsOutbound.forEach((_, i) => {
            graph.emitTo(panel, i, targets.includes(i) ? signal : null)
        })

        panel.state = signal ? 'routing' : 'off'

        const prevKey = JSON.stringify(prev)
        const nextKey = JSON.stringify(chosen)
        if (prevKey !== nextKey)
            DecisionNode.dispatch(panel, 'decision:routed', { from: prev, to: chosen, input: inputIndex })

        graph.updateAllGenDraws()
    }

    // ── Override this in subclasses ───────────────────────────────────────────

    /**
     * Return the output index (or array of indices) to route the signal to.
     * Return `null` to block all outputs.
     *
     * @param  {Object}      panel       — reactive panel state
     * @param  {Object|null} signal      — { v, a } or null — signal on the triggering pip
     * @param  {number}      inputIndex  — which inbound pip triggered this call
     * @param  {PowerGraph}  graph       — graph engine (rarely needed)
     * @returns {number | number[] | null}
     *
     * Access all inbound pip signals via panel._pipSignals:
     *   panel._pipSignals[0]  — last signal seen on inbound pip 0
     *   panel._pipSignals[1]  — last signal seen on inbound pip 1
     */
    static decide(panel, signal, inputIndex, graph) {
        if (typeof panel.decideCallback === 'function') {
            return panel.decideCallback(signal, inputIndex, panel, graph) ?? 0
        }
        return panel.defaultOutput ?? 0
    }

    /**
     * Null every outbound pip when the node is disabled, not just pip 0.
     */
    static onDisabled(panel, graph) {
        panel.pipsOutbound.forEach((_, i) => graph.emitTo(panel, i, null))
    }

    // ── Actions ───────────────────────────────────────────────────────────────

    /** Force a manual re-route from the UI. */
    static reRoute(panel, graph) {
        panel.pipsOutbound.forEach((_, i) => graph.emitTo(panel, i, null))
        this._route(panel, panel.signal, graph, panel._lastInPip ?? 0)
    }

    /** Pin a default output index from the UI. */
    static setDefault(panel, index, graph) {
        panel.defaultOutput = index
        this._route(panel, panel.signal, graph, panel._lastInPip ?? 0)
    }
}

NodeRegistry.register(DecisionNode)

class MyCustomDecision extends DecisionNode {

    static type        = 'my-decision'
    static label       = 'My Decision'
    static outputCount = 3

    static catalog = [
        { key: 'my-decision', label: 'My Decision (3 out)', outputCount: 3, inputCount: 2 },
    ]

    /**
     * Randomly routes to one of the available outputs each time a signal arrives
     * or the tick interval fires. panel._pipSignals[n] holds per-input history.
     */
    static decide(panel, signal, inputIndex, graph) {
        if (!signal) return null
        // Return a random output for demo purposes
        return Math.floor(Math.random() * panel.outputCount)
    }
}

NodeRegistry.register(MyCustomDecision)

File: power_graph/src/power_graph/nodes/decision.py
"""
nodes/decision.py
──────────────────────────────────────────────────────────────────────────────
Decision — multi-output routing node.

States: 'off' | 'routing' | 'blocked' | 'error'

Override `decide(panel, signal, input_index, graph)` in a subclass to
implement custom routing logic.  The method should return:
  • an int  — single output pip index to receive the signal
  • a list  — multiple output pip indices each receive the signal
  • None    — block all outputs

Default routing: input pip at index N → output pip at index N (1:1).

Per-pip signal cache (`_pip_signals`) allows the node to track what arrived
on each inbound pip for the re-routing tick.

Example subclass:

    class Toggle2Way(Decision):
        type  = 'toggle-2way'
        label = 'Toggle 2-Way'
        catalog = [{'key': 'tog2', 'label': 'Toggle 2-Way',
                    'inputCount': 1, 'outputCount': 2}]

        _side = 0  # class-level for simplicity; use panel dict in practice

        @classmethod
        def decide(cls, panel, signal, input_index, graph):
            cls._side ^= 1
            return cls._side

    NodeRegistry.register(Toggle2Way)
"""

from typing import Dict, List, Optional, Union
from power_graph.node_base import NodeBase, Signal
from power_graph.node_registry import NodeRegistry


class Decision(NodeBase):
    """Multi-output routing node with override-able decision logic."""

    type  = 'decision'
    label = 'Decision'
    group = 'Routing'
    dispatch_delay = 50

    catalog = [
        {'key': 'decision-1-2', 'label': '1→2 Router',  'inputCount': 1, 'outputCount': 2},
        {'key': 'decision-2-2', 'label': '2→2 Router',  'inputCount': 2, 'outputCount': 2},
        {'key': 'decision-1-4', 'label': '1→4 Router',  'inputCount': 1, 'outputCount': 4},
    ]

    @classmethod
    def _default_pips_inbound(cls, node_id: int) -> List[Dict]:
        return [{'label': f'{node_id}:in0', 'index': 0}]

    @classmethod
    def _default_pips_outbound(cls, node_id: int) -> List[Dict]:
        return [
            {'label': f'{node_id}:out0', 'index': 0},
            {'label': f'{node_id}:out1', 'index': 1},
        ]

    @classmethod
    def defaults(cls, node_id: int, preset: Dict = None) -> Dict:
        if preset is None:
            preset = {}
        in_count  = preset.get('inputCount', 1)
        out_count = preset.get('outputCount', 2)
        base = super().defaults(node_id, preset)
        base['pipsInbound']  = [{'label': f'{node_id}:in{i}',  'index': i} for i in range(in_count)]
        base['pipsOutbound'] = [{'label': f'{node_id}:out{i}', 'index': i} for i in range(out_count)]
        base.update({
            'inputCount':    in_count,
            'outputCount':   out_count,
            'tickInterval':  preset.get('tickInterval', 0.0),
            'lastDecision':  None,
            '_tick_accum':   0.0,
            '_pip_signals':  {},      # {pip_index: signal}
            'state':         'off',
        })
        return base

    @classmethod
    def config_fields(cls) -> List[str]:
        return [*super().config_fields(), 'inputCount', 'outputCount', 'tickInterval']

    # ── Override this ─────────────────────────────────────────────────────────

    @classmethod
    def decide(cls, panel: Dict, signal: Signal, input_index: int,
               graph) -> Optional[Union[int, List[int]]]:
        """
        Return the output pip index (or list of indices) to route signal to.
        Return None to block all outputs.

        Default: route to the matching output index.
        """
        out_count = panel.get('outputCount', 2)
        idx       = input_index % out_count
        return idx if idx < out_count else None

    # ── Core logic ────────────────────────────────────────────────────────────

    @classmethod
    def apply(cls, panel: Dict, signal: Signal, graph):
        # Determine which inbound pip this signal arrived on.
        # Graph passes inbound pip index via panel's last received pip (if tracked).
        # For simplicity we use pip index 0 by default; subclasses override decide().
        in_idx = panel.get('_last_in_pip', 0)
        panel['_pip_signals'][in_idx] = signal

        if signal is None and all(s is None for s in panel['_pip_signals'].values()):
            if panel['state'] != 'off':
                cls.dispatch(panel, 'state:change', {'from': panel['state'], 'to': 'off'})
            panel['state'] = 'off'

        cls._route(panel, signal, in_idx, graph)

    @classmethod
    def tick(cls, panel: Dict, dt: float, graph):
        interval = panel.get('tickInterval', 0.0)
        if not interval:
            return
        panel['_tick_accum'] = panel.get('_tick_accum', 0.0) + dt
        if panel['_tick_accum'] >= interval:
            panel['_tick_accum'] = 0.0
            for in_idx, sig in panel.get('_pip_signals', {}).items():
                cls._route(panel, sig, in_idx, graph)

    @classmethod
    def _route(cls, panel: Dict, signal: Signal, in_idx: int, graph):
        """Apply decide() and emit to chosen output(s)."""
        out_count = panel.get('outputCount', 2)
        result    = cls.decide(panel, signal, in_idx, graph)

        if result is None:
            # Block all
            chosen = []
            new_state = 'blocked'
        elif isinstance(result, int):
            chosen    = [result]
            new_state = 'routing' if signal is not None else 'off'
        else:
            chosen    = list(result)
            new_state = 'routing' if signal is not None else 'off'

        panel['lastDecision'] = chosen

        if panel['state'] != new_state:
            cls.dispatch(panel, 'state:change', {'from': panel['state'], 'to': new_state})
        panel['state'] = new_state

        for out_idx in range(out_count):
            out_signal = signal if out_idx in chosen else None
            graph.emit_pip(panel, out_idx, out_signal)

    @classmethod
    def reset(cls, panel: Dict, graph):
        prev = panel['state']
        panel.update({'state': 'off', '_pip_signals': {}, 'lastDecision': None,
                      '_tick_accum': 0.0})
        cls.dispatch(panel, 'state:change', {'from': prev, 'to': 'off'})
        for i in range(panel.get('outputCount', 2)):
            graph.emit_pip(panel, i, None)


NodeRegistry.register(Decision)
